Cmake build fixes

hayden-donnelly opened the pull request:
I originally developed the python package alongside a nix derivation (`./moondream-ggml/moondream-ggml-python.nix`) for installation. The nix installation worked but I never tested it with pip. Today I tried to install with pip for the first time and I ran into some issues. This fixes them. I'm pretty sure I broke the nix installation though.

The main problem was non position independent code. Since `libggml` and `libmoondream_ggml` were built as static libraries, they contained position dependent code which cause problems when they were linked to the `cpp_ffi` shared library. I added an option for building `libggml` and `libmoondream_ggml` as shared libraries in order to resolve this. I'm not sure why this was only an issue with the pip install. Maybe because of different gcc versions?

I also encountered some problems with SIMD support because I did all of this in a VirtualBox VM. In my case the solution was manually setting `-DGGML_FMA=ON` but there will likely be different problems/solutions in different environments. We'll probably need some additional configuration code in `./moondream-ggml/bindings/python/setup.py` in order to detect architecture/SIMD support and set the appropriate GGML build flags.

Commit:
add option to make all libs shared

This fixes non position independent code errors
when linking static libraries to shared libraries.
This is required when building a python package
because the final form of the C++ extension module
is a shared library.

- Added option to toggle executable build so that
  the cmake build will yield either an executable
  target or a library target, but not both.
- Added option to toggle shared library build.
  This is used in ./bindings/python/setup.py.
- Fixed python package shared library loading
  outside of a nix derivation. This means that
  the package should now work properly when
  installed with pip.

Commit:
update moondream-ggml readme for cmake changes

vikhyat commented:
@hayden-donnelly looks like this PR is in draft status, was that intentional?

cc @jasonallen

hayden-donnelly commented:
@vikhyat yeah I'm gonna see if I can fix the Metal error and add it to the PR

Commit:
disable metal when building ggml

This bypasses the cmake read-only filesystem error and allows the program build,
but it doesn't solve the root of the issue.

Commit:
remove Metal flag override and set cmake runtime out dir

This also fixes the cmake read-only filesystem error for Metal but introduces
a compiler error:

[  5%] Building C object dependencies/ggml/src/CMakeFiles/ggml.dir/ggml-metal.m.o
In file included from /Users/hayden/Desktop/repos/moondream/moondream-ggml/core/dependencies/ggml/src/ggml-metal.m:6:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:12:
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSBundle.h:91:143: error: function does not return NSString
- (NSAttributedString *)localizedAttributedStringForKey:(NSString *)key value:(nullable NSString *)value table:(nullable NSString *)tableName NS_FORMAT_ARGUMENT(1) NS_REFINED_FOR_SWIFT API_AVAILABLE(macos(12.0), ios(15.0), watchos(8.0), tvos(15.0));
                                                         ~~~~~~~~~~~~~~                                                                       ^                  ~
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:103:48: note: expanded from macro 'NS_FORMAT_ARGUMENT'
        #define NS_FORMAT_ARGUMENT(A) __attribute__ ((format_arg(A)))

I'm guessing this is a problem with my version of CommandLineTools (too old)
and won't replicate on newer versions.

hayden-donnelly commented:
Two solutions to the read-only file system error:

[bdbd8b5](https://github.com/vikhyat/moondream/pull/130/commits/bdbd8b58de551cdd9ab4b69532a0cc1d5d03ee79) simply excludes Metal support regardless of platform. This removes the error and lets me build everything normally.

[7c2853a](https://github.com/vikhyat/moondream/pull/130/commits/7c2853a9023533874150eeacfe4009ce33f9e3b3) includes Metal support on Apple platforms (the default GGML behavior) and `CMAKE_RUNTIME_OUTPUT_DIRECTORY`. This removes the cmake errors but introduces a compilation error for `ggm-metal.m`. I think this is because the macbook I'm using has an old version of CommandLineTools (`12.5.1.0.1.1623191612`), so it might work if you guys have a more current setup.
